pkg/graveler/committed/apply.go

```diff
@@ -26,99 +26,129 @@ func incrementDiffSummary(d *graveler.DiffSummary, typ graveler.DiffType) {
+			} else {
+				if logger.IsTracing() {
+					logger.WithFields(logging.Fields{
+						"from": string(iterRange.MinKey),
+						"to":   string(iterRange.MaxKey),
+						"ID":   iterRange.ID,
+					}).Trace("copy entire range at end")
+				}
+				if err := writer.WriteRange(*iterRange); err != nil {
+					return 0, fmt.Errorf("copy iter range %s: %w", iterRange.ID, err)
+				}
+				count += int(iterRange.Count)
 			}
-			if !source.NextRange() {
+			if !iter.NextRange() {
 				break
 			}
 		} else {
 			if logger.IsTracing() {
+				if iterValue.IsTombstone() {
+					// internal error but no data lost: deletion requested of a
+					// file that was not there.
+					logger.WithField("id", string(iterValue.Identity)).Warn("[I] unmatched delete")
+					continue
+				}
 				logger.WithFields(logging.Fields{
-					"key": string(sourceValue.Key),
-					"ID":  string(sourceValue.Identity),
-				}).Trace("write key from source at end")
+					"key": string(iterValue.Key),
+					"ID":  string(iterValue.Identity),
+				}).Trace("write key from iter at end")
 			}
-			if err := writer.WriteRecord(*sourceValue); err != nil {
-				return fmt.Errorf("write source record: %w", err)
+			if err := writer.WriteRecord(*iterValue); err != nil {
+				return 0, fmt.Errorf("write iter record: %w", err)
 			}
-			if !source.Next() {
+			count++
+			if !iter.Next() {
 				break
 			}
 		}
 	}
-	return source.Err()
+	return count, iter.Err()
 }
 
-// applyFromDiffs applies all changes from diffs to writer and returns the number elements it
-// added.
-func applyFromDiffs(ctx context.Context, logger logging.Logger, writer MetaRangeWriter, diffs graveler.ValueIterator) (int, error) {
-	numAdded := 0
-	for {
-		select {
-		case <-ctx.Done():
-			return 0, ctx.Err()
-		default:
-		}
-
-		diffValue, haveDiffs := diffs.Value(), diffs.Next()
-		if diffValue.IsTombstone() {
-			// internal error but no data lost: deletion requested of a
-			// file that was not there.
-			logger.WithField("id", string(diffValue.Identity)).Warn("[I] unmatched delete")
-			continue
-		}
-		if logger.IsTracing() {
-			logger.WithFields(logging.Fields{
-				"key":       string(diffValue.Key),
-				"ID":        string(diffValue.Identity),
-				"tombstone": diffValue.IsTombstone(),
-			}).Trace("write key from diffs at end")
-		}
-		if err := writer.WriteRecord(*diffValue); err != nil {
-			return 0, fmt.Errorf("write added record: %w", err)
-		}
-		numAdded++
-		if !haveDiffs {
-			break
+func hasChanges(summary graveler.DiffSummary) bool {
+	for _, changes := range summary.Count {
+		if changes > 0 {
+			return true
 		}
 	}
-	return numAdded, nil
+	return false
 }
 
-func Apply(ctx context.Context, writer MetaRangeWriter, source Iterator, diffs graveler.ValueIterator, opts *ApplyOptions) (graveler.DiffSummary, error) {
+func Apply(ctx context.Context, writer MetaRangeWriter, source Iterator, diffs Iterator, opts *ApplyOptions) (graveler.DiffSummary, error) {
 	logger := logging.FromContext(ctx)
```

Comment:
Note that this "cyclomatic complexity" nonsense already bit us, which is why I split off the 2 functions at the top.  We should probably bite off a bit more here, maybe stuff like keeping `haveSource`, `haveDiffs` inside a state object with the iterators and performing actions on _that_; or just passing that state object into a separate function that holds each of the `select`s.